pupil_src/shared_modules/surface_tracker/surface_tracker.py

```diff
@@ -0,0 +1,458 @@
+                min_marker_perimeter=self.marker_min_perimeter,
+                invert_image=self.inverted_markers,
+            )
+        else:
+            markers = marker_det.detect_markers(
+                gray,
+                grid_size=5,
+                aperture=11,
+                min_marker_perimeter=self.marker_min_perimeter,
+            )
+
+        # Robust marker detection requires previous markers to be in a different format than the surface tracker.
+        self.markers_dict = markers
+        markers = [
+            Square_Marker_Detection(
+                m["id"], m["id_confidence"], m["verts"], m["perimeter"]
+            )
+            for m in markers
+        ]
+        self.markers_unfiltered = markers
+        self.markers = self._filter_markers(markers)
+
+    def _filter_markers(self, markers):
+        filtered_markers = [
+            m
+            for m in markers
+            if m.perimeter >= self.marker_min_perimeter
+            and m.id_confidence > self.marker_min_confidence
+        ]
+
+        # if an id shows twice use the bigger marker (usually this is a screen camera echo artifact.)
+        marker_by_id = {}
+        for m in filtered_markers:
+            if m.id not in marker_by_id or m.perimeter > marker_by_id[m.id].perimeter:
+                marker_by_id[m.id] = m
+
+        return list(marker_by_id.values())
+
+    @abstractmethod
+    def _update_surface_locations(self, idx):
+        pass
+
+    @abstractmethod
+    def _update_surface_corners(self):
+        pass
+
+    def _add_surface_events(self, events, frame):
+        """
+        Adds surface events to the current list of events.
+
+        Args:
+            events: Current list of events.
+            frame: The according world camera frame
+        """
+        events["surfaces"] = []
+        for surface in self.surfaces:
+            if surface.detected:
+                gaze_events = events.get("gaze", [])
+                gaze_on_surf = surface.map_events(gaze_events, self.camera_model)
+                fixation_events = events.get("fixations", [])
+                fixations_on_surf = surface.map_events(
+                    fixation_events, self.camera_model
+                )
+
+                surface_event = {
+                    "topic": "surfaces.{}".format(surface.name),
+                    "name": surface.name,
+                    "surf_to_img_trans": surface.surf_to_img_trans.tolist(),
+                    "img_to_surf_trans": surface.img_to_surf_trans.tolist(),
+                    "gaze_on_surf": gaze_on_surf,
+                    "fixations_on_surf": fixations_on_surf,
+                    "timestamp": frame.timestamp,
+                }
+                events["surfaces"].append(surface_event)
+
+    @abstractmethod
+    def _update_surface_heatmaps(self):
+        pass
+
+    def gl_display(self):
+        self.gui.update()
+
+    def add_surface(self, _=None, init_dict=None):
```

Comment:
This function is used as callback function for the `add surface thumb`. The API for those callback functions requires them to accept an argument. This is my way of handling this. Do you think it would make more sense to use a named parameter which is never used?